app/src/src_nonfreenet/org/breezyweather/sources/lvgmc/LvgmcService.kt

```diff
@@ -0,0 +1,291 @@
+        val airQualityLocation = location.parameters.getOrElse(id) { null }?.getOrElse("airQualityLocation") { null }
+
+        if (currentLocation.isNullOrEmpty() || airQualityLocation.isNullOrEmpty()) {
+            return Observable.error(InvalidLocationException())
+        }
+
+        val failedFeatures = mutableListOf<SourceFeature>()
+
+        val current = if (requestedFeatures.contains(SourceFeature.FEATURE_CURRENT)) {
+            mApi.getCurrent().onErrorResumeNext {
+                failedFeatures.add(SourceFeature.FEATURE_CURRENT)
+                Observable.just(emptyList())
+            }
+        } else {
+            Observable.just(emptyList())
+        }
+
+        val formatter = SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.ENGLISH)
+        formatter.timeZone = TimeZone.getTimeZone("Europe/Riga")
+        val fromDate = formatter.format(Date(Date().time - 24.hours.inWholeMilliseconds))
+        val airQuality = if (requestedFeatures.contains(SourceFeature.FEATURE_AIR_QUALITY)) {
+            mApi.getAirQuality(
+                station = airQualityLocation,
+                fromDate = fromDate
+            ).onErrorResumeNext {
+                failedFeatures.add(SourceFeature.FEATURE_AIR_QUALITY)
+                Observable.just(emptyList())
+            }
+        } else {
+            Observable.just(emptyList())
+        }
+
+        return Observable.zip(current, airQuality) {
+                currentResult: List<LvgmcCurrentResult>,
+                airQualityResult: List<LvgmcAirQualityResult>,
+            ->
+            convertSecondary(
+                location = location,
+                currentResult = if (requestedFeatures.contains(SourceFeature.FEATURE_CURRENT)) {
+                    currentResult
+                } else {
+                    null
+                },
+                airQualityResult = if (requestedFeatures.contains(SourceFeature.FEATURE_AIR_QUALITY)) {
+                    airQualityResult
+                } else {
+                    null
+                },
+                failedFeatures = failedFeatures
+            )
+        }
+    }
+
+    // REVERSE GEOCODING
+    override fun requestReverseGeocodingLocation(
+        context: Context,
+        location: Location,
+    ): Observable<List<Location>> {
+        val formatter = SimpleDateFormat("yyyyMMddHH", Locale.ENGLISH)
+        formatter.timeZone = TimeZone.getTimeZone("Europe/Riga")
+        val laiks = formatter.format(Date()) + "00"
+        val bounds = "POLYGON((" +
+            (location.longitude - 0.1) + " " + (location.latitude - 0.1) + ", " +
+            (location.longitude + 0.1) + " " + (location.latitude - 0.1) + ", " +
+            (location.longitude + 0.1) + " " + (location.latitude + 0.1) + ", " +
+            (location.longitude - 0.1) + " " + (location.latitude + 0.1) + ", " +
+            (location.longitude - 0.1) + " " + (location.latitude - 0.1) + "))"
```

Comment:
Oh gosh

Reply:
I will move this into a separate function.